### ISSUE TITLE
Fire hazard error handling help please. 🔥 

### ISSUE THREAD

**Issue by RIAEvangelist** [chrisJohn404/LabJack-nodejs#19]
I get error 1298 and 1239 a lot. Both are driver operation and reconnect errors.

I am using a t4 and a t7 in battery cell cycling software. When the controls fail, it can lead to a fire if not caught. We have backup safeties in place, but those will only prevent the actual fire, they will not prevent damage to the cell due to the labjack driver crashing.

Is there a way I could handle this in software to prevent our cycling rig from catching on fire and burning the building down?

Thanks in advance for any help.

**Comment by chrisJohn404** [comment]
What host computer operating system and communication interface are you using?

As LabJack-nodejs is a wrapper for the LJM driver, error codes 1298 and 1239 are LJME_ATTR_LOAD_COMM_FAILURE and LJME_RECONNECT_FAILED respectively [1].

In some situations, typically for the T7-Pro and connecting over WiFi, error code LJME_ATTR_LOAD_COMM_FAILURE (1298) can happen when initially connecting to a device during the "open" call or during a device discovery operation[2].  To address this issue you may need to loop on the open call until the connection attempt is successful.  You should also consider moving the device closer to the WiFi access point or away from RF interference in the area.  LJME_ATTR_LOAD_COMM_FAILURE (1239) occurs during the device connection establishment step when it's name among other settings core to the LJM Library's functionality are being read but an error is encountered.

During typical operation these error codes are reported after previously connecting to and communicating with a device.  LJME_RECONNECT_FAILED (1239) indicates that the communication interface between the host computer and the slave device (LabJack) has failed.  This can happen for numerous reasons, sometimes indicating that an electrostatic discharge (ESD) event has occurred; illuminating one of the best reasons to use a LabJack, it can be the sacrificial part in your system instead of your computer.  Be sure to notify your hardware engineer of these situations however as over time a host PC's hardware may become degraded causing system wide failures.  A few examples what may cause the LJME_RECONNECT_FAILED (1239) error code are as follows:

1. You are connected to the device over USB or Ethernet and the device experiences an ESD shock that disrupts communication but does not break the device.  Read or write attempts from your program will report a LJME_RECONNECT_FAILED (1239) error due to the device I/O attempt timing out.  Depending on your LJM configurations and your program you may get this error repeatedly until the PC and device negotiate a connection again.
2. You are connected over WiFi and there is an RF blast that causes the device's connection to become interrupted.  The device will attempt to connect to the configured access point and establish a new connection.  In edge cases a LJME_ATTR_LOAD_COMM_FAILURE (1239) may occur indicating that the connection type is having troubles becoming established once again however the typical error, if you continue to perform device I/O will be LJME_RECONNECT_FAILED (1239).
3. You are connected to the device via any connection type and one of the features of the device causes a watchdog timeout to occur indicating that the device has encountered an unrecoverable error and needs to restart.  The device will restart and have the hardware/software stacks will establish communication again causing the LJME_RECONNECT_FAILED (1239) error.

Thanks to the LJM library and how the LabJack device is configured, the host PC and the LabJack will typically be able to re-establish the communication over a previously negotiated connection if disrupted for some reason.  All you need to do is continue to read any register [3] from the device until the error code stops occurring.  In some extreme cases you may need to use the device opening commands to establish communication with a secondary communication interface to recover communications through the primary communication interface.

1. [LJM Error Codes](https://labjack.com/support/software/api/ljm/error-codes).
2. Device discovery operation or [device information function](https://labjack.com/support/software/api/ljm/function-reference/device-information)
3. Modbus [register](https://labjack.com/support/software/api/modbus/modbus-map)

**Comment by RIAEvangelist** [comment]
@chrisJohn404 this is a fresh and updated version of Windows 10. Connecting via USB.

If the issue happened during connect this would not be as big an issue. As it stands, it happens during the middle of our test on the T4 and the T7 Pro. When the labjack communication fails, it is stuck in the state when it failed. This is an issue when the heaters are turned on as it can cause a thermal runaway of both the heating elements and the battery itself leading to not just a failed test, but also a fire. The failed test alone is a large enough reason for us to have started considering that we need to completely rebuild our system without a labjack, but the fire hazard and inability to recover is pushing the stakeholders to look for new hardware.  The labjack is fully isolated from the computer via isolated USB connections.

I believe our issue is as described in your 1. description. I can not confirm if it is an ESD issue though. The reason I think this is because if I disable and reenable the device, via the device manager, I am able to connect to the labjack again without power cycling it. The same issue is present in killing and testing shows the same solution works there.

LAst friday I was able to incorporate a real time websocket server into our application to view real time thermocouple information, and I noticed that there were some areas with loss of communication for multiple minutes. This is an even larger problem because it means our test information could be dramatically flawed and still result in a fire.

Short of replacing the labjack hardware and completely recoding our software stack to use something the stakeholders can have confidence in, is there anything I can do to make this issue more transparent and resolvable? I really do not want to have to recode ***everything*** as this is months of work. I would much rather find a way to handle the issues in real time so we can keep the labjacks as a core part of our system.

Any advice or code snippets you can provide as an example would be greatly appreciated. I will also speak with our hardware engineer to see if we can detect any ESD related issues.

I look forward to anything you can suggest. Thank you.

**Comment by chrisJohn404** [comment]
For the LJME_ATTR_LOAD_COMM_FAILURE (1239), I should have also linked to the LJM user's guide page [5.16 - LJME_ATTR_LOAD_COMM_FAILURE](https://labjack.com/support/software/api/ljm/ljmeattrloadcommfailure).  I will edit it.

What firmware versions are you using?  For debugging purposes, please update to the latest stable firmware using Kipling's [device updater](https://labjack.com/support/software/applications/t-series/kipling/device-updater) page.

> The reason I think this is because if I disable and reenable the device, via the device manager, I am able to connect to the labjack again without power cycling it.

Given this statement you have been able to reproduce this issue.  At the time of failure do the device's LEDs show normal behavior?

**Comment by RIAEvangelist** [comment]
We are on the latest firmware, just updated via Kipling last week sometime.

As far as reproducing the issue, it happens often enough as to make the stakeholders consider paying me to redo months worth of work to fix the issue with different hardware because of the strong possibility of ad data and of course the whole fire hazard thing. This isn't really a once a month thing, it's more like once every day or every other day.

Also, yes, all of the LEDs seem to operate properly still.

**Comment by chrisJohn404** [comment]
> We are on the latest firmware, just updated via Kipling last week sometime ... yes, all of the LEDs seem to operate properly still.

It is really important to note whether or not there is any "Combined LED Activity" while your tests are being performed.

Can you post a list of the registers you are reading from and writing to?  To control things it's possible to be writing to a FIO/EIO/CIO/DIO or a related DIO_EF register, one of the DAC registers, a USER_RAM register, or a _DEFAULT register.

Can you confirm whether or not you are reading from an [AIN_EF register](https://labjack.com/support/datasheets/t-series/ain/extended-features/thermocouple) to retrieve thermocouple readings and also tell me if you are using any LabJack accessories?

**Comment by RIAEvangelist** [comment]
Sure thing, here is our T4 config which we have a relay board attachment on.

These are all DIO.

```js

  
import {t4,readT4Value,writeT4Value} from '../utils/readWrite.js';

const emergency={
    
}

const safety={
    //emergency disconnect for cell and heaters
    cellAndHeaterCoil:{
        'type':'DIO',
        'signal':18,
        'faultState':1
    },
    ACMeanwellHeaterPowerEnable:{
        'type':'DIO',
        'signal':8,
        'faultState':1
    },
    heaterAutoPowerEnable:{
        'type':'DIO',
        'signal':17,
        'faultState':1
    }
}

const heating={
    heaterCanCrimp:{
        'type':'DIO',
        'signal':16,
        'faultState':1
    },
    heaterCanBase:{
        'type':'DIO',
        'signal':15,
        'faultState':1
    }
};

const cooling={
    coolantFlow:{
        'type':'DIO',
        'signal':11,
        'faultState':0
    },
    coolantFlowValve:{
        'type':'DIO',
        'signal':13,
        'faultState':0
    },
    coolantBypassRecirculationValve:{
        'type':'DIO',
        'signal':14,
        'faultState':1
    }
}

const vacuum={
    ACPump:{
        'type':'DIO',
        'signal':12,
        'faultState':0
    }
}

const TECCoolers={
    coolerBase:{
        'type':'DIO',
        'signal':9,
        'faultState':1
    },
    coolerCrimp:{
        'type':'DIO',
        'signal':10,
        'faultState':1
    }
}

const logAllValues=function(){
    console.log('logging all control configs');
    for(const [key,value] of Object.entries(configs)){
        for(const [key,config] of Object.entries(value)){
            console.log(`${key} (${config.type}${config.signal}) : ${config.value}`);
        }
    } 
}

const populateValue=function(key,config){
    //console.log(`setting up getter for ${key}`);
    Object.defineProperty(
        config, 
        'value', 
        { 
            get: lastSetValue.bind(config),
            set: writeT4Value.bind(config)
        }
    );
    //console.log(config);
}
const lastSetValue=function(){
    return this._lastSetValue
}

const configs={emergency,safety,heating,cooling,vacuum,TECCoolers}

for(const [key,value] of Object.entries(configs)){
    for(const [key,config] of Object.entries(value)){
        populateValue(key,config);
    }
}



const old={
    configs,t4,logAllValues
}

export { 
    old as default, 
    emergency,
    safety,
    heating,
    cooling,
    vacuum,
    TECCoolers,
    t4,
    logAllValues
};  

```

And here is our t7 pro config which we have an expansion board on. 

These are all AIN(signal)

```js

import {t7,readT7Value,writeT7Value} from '../utils/readWrite.js';

const thermocouples={
    cellCrimpSide:{
        'signal':48
    },
    cellBaseSide:{
        'signal':49
    },
    can30mm:{
        'signal':50
    },
    can10mm:{
        'signal':51
    },
    can40mm:{
        'signal':52
    },
    can60mm:{
        'signal':53
    },
    aluminumFixtureCellCrimp:{
        'signal':54
    },
    aluminumFixtureCellBase:{
        'signal':55
    },
    liquidCoolantIn:{
        'signal':80
    },
    liquidCoolantOut:{
        'signal':81
    },
    // TECColdBase:{
    //     'signal':82
    // },
    TECColdCrimp:{
        'signal':83
    },
    TECColdMiddle:{
        'signal':84
    },
    heaterCanBase:{
        'signal':85
    },
    heaterCanCrimp:{
        'signal':86
    }
}

const logAllValues=function(){
    for(const [key,config] of Object.entries(thermocouples)){
        const value=config.value;
        console.log(`${key} (AIN${config.signal}) ${
            value.degC.toFixed(2)
        } C `);
    }    
}

const populateValue=function(key,config){
    //console.log(`setting up getter for ${key}`);
    Object.defineProperty(
        config, 
        'value', 
        { 
            get: readT7Value.bind(config)
        }
    );
    //console.log(config);
}

const init=function(){
    for(let [id,thermocouple] of Object.entries(thermocouples)){
        const signal=`AIN${thermocouple.signal}`;
        
        t7.writeSync(`${signal}_EF_INDEX`, 24);
        //console.log(`${signal}_EF_INDEX is set to 24 or type T thermo couple`);
        
        t7.writeSync(`${signal}_EF_CONFIG_A`, 1);
        //console.log(`${signal}_EF_CONFIG_A is set to 1 reading degC`);
    
        t7.writeSync(`${signal}_EF_CONFIG_B`, 60052);
        // console.log(`${signal}_EF_CONFIG_B is set to ${gnd} `);
    }
}

for(const [key,config] of Object.entries(thermocouples)){
    populateValue(key,config);
}

const old={
    thermocouples,t7,logAllValues,init
}

export { old as default, thermocouples,t7,logAllValues,init};

```

**Comment by RIAEvangelist** [comment]
And also, just to parse out the part about the EF registry, for the thermo couples on the t7pro I am writing to these EF_registries upon connect. This is all in the above snippet, I just added it here to make it easier for you to see.

As for combined LED activity, until the unhandled LJ error causes my code to crash, there is a lot of LED activity. Once the code crashes, I can not confirm if there is any further activity, like delayed commands being sent from the windows driver or whatnot. Right now, I remote into the work station and code from there with a webcam pointing at the device.

```js

const init=function(){
    for(let [id,thermocouple] of Object.entries(thermocouples)){
        const signal=`AIN${thermocouple.signal}`;
        
        t7.writeSync(`${signal}_EF_INDEX`, 24);
        //console.log(`${signal}_EF_INDEX is set to 24 or type T thermo couple`);
        
        t7.writeSync(`${signal}_EF_CONFIG_A`, 1);
        //console.log(`${signal}_EF_CONFIG_A is set to 1 reading degC`);
    
        t7.writeSync(`${signal}_EF_CONFIG_B`, 60052);
        // console.log(`${signal}_EF_CONFIG_B is set to ${gnd} `);
    }
}

```

[t4 relay board 1](https://i.imgur.com/M4e3nIk.jpg)
[t4 relay board 2](https://i.imgur.com/f8B0BlD.jpg)
[t4 relay board hooked up](https://i.imgur.com/E0MhlOM.jpg)
[t7 pro with expansion hookedup for thermocouples](https://i.imgur.com/PSUxtHR.jpg)
[remote view](https://i.imgur.com/d8e6u9K.jpg)
[data samples](https://i.imgur.com/LPhAPZA.jpg)

We are measuring temps ranging from -100deg C to 100deg C in this vacuum chamber.

**Comment by RIAEvangelist** [comment]
Also, here is a screen shot of just the t7 running and failing. This happens multiple times a day when monitoring temperatures.

I just disable the device via the device manager and reenable it and it starts working again.

![dead t7](https://i.imgur.com/jLbxwGg.png)

**Comment by RIAEvangelist** [comment]
I would also like to handle the `DriverOperationError`, is there any specific way to handle this or do I just need to listen for otherwise unhandled errors in the process?

**Comment by RIAEvangelist** [comment]
One more code snippit, I am indeed reading the `AIN_EF_READ_A`


```js

const readT7Value=function(){
    this.voltage=t7.readSync(`AIN${this.signal}`).toFixed(5);
    this.rawTC=Number(
        t7.readSync(`AIN${this.signal}_EF_READ_A`)
    );

    this.degC=getAdjustedTemp(this.rawTC);
    //this.degC=this.rawTC;

    return this;
}

```

**Comment by RIAEvangelist** [comment]
Can I just throw a `try` `catch` around the read?

**Comment by chrisJohn404** [comment]
You may indeed be experiencing an electrical transient (EMI or ESD) event when turning on or off one of your relays that is causing your device to stop communicating with your PC through the USB connection interface.  Aside from there possibly being some [ground loops]() that could be removed, you can try some of the suggestions in the [USB Communication Failure](https://labjack.com/support/app-notes/usb-communication-failure) AppNote.  Look specifically at the answer section for C & D:

> Remove all connections from the device, except for USB, and see if the problem goes away.  If the problem does not go away, try a different location, computer, and USB cable.  If the problem does go away, start adding back connections and see if you can isolate a particular connection that causes the problem, and then you can consider adding extra protection (chokes/ferrites/inductors, series resistors, parallel capacitors & TVS diodes) to problem lines.  We found that a 1mH inductor in series with each line can be useful, but you have to do this on all lines including those to GND.

How are the two devices connected to your PC, do you just have the two USB cables going straight into two USB ports on a desktop PC?  This may not be the best solution as it may punt the issue down the road but you could try adding two powered USB hubs, one for each device.

> const thermocouples={
    cellCrimpSide:{
        'signal':48
    },
    cellBaseSide:{
        'signal':49
    },

Regarding your first code sample where you are configuring the T7 to read thermocouples, make sure you are following the note on the [T7 Thermocouple AppNote](https://labjack.com/support/app-notes/thermocouples/t7) and to use differential analog inputs and not single ended ones.  A few years back LabJack started changing their documentation to indicate that the T7 supports reading from up to 42 thermocouples and not one per channel resulting in 84.

> DriverOperationError

To catch these errors, yes, wrap device I/O calls with [try - catch](https://nodejs.org/en/knowledge/errors/what-is-try-catch/) statements.

**Comment by RIAEvangelist** [comment]
> You may indeed be experiencing an electrical transient (EMI or ESD) event when turning on or off one of your relays that is causing your device to stop communicating with your PC through the USB connection interface. Aside from there possibly being some ground loops that could be removed, you can try some of the suggestions in the USB Communication Failure AppNote. Look specifically at the answer section for C & D:

The same issue arises when the entire T4 is removed and we only monitor temperatures via the t7 pro `AIN`

> How are the two devices connected to your PC, do you just have the two USB cables going straight into two USB ports on a desktop PC? This may not be the best solution as it may punt the issue down the road but you could try adding two powered USB hubs, one for each device.

We have 2 isolated usbs with separate power supplies

> Regarding your first code sample where you are configuring the T7 to read thermocouples, make sure you are following the note on the T7 Thermocouple AppNote and to use differential analog inputs and not single ended ones. A few years back LabJack started changing their documentation to indicate that the T7 supports reading from up to 42 thermocouples and not one per channel resulting in 84.

These are all differential

> To catch these errors, yes, wrap device I/O calls with try - catch statements.

I'll give this a shot and see if I can make the labjack work well enough to be able to keep using it.

**Comment by RIAEvangelist** [comment]
It took some redoing because I wanted to handle sync issues with sync delays, and that was not so simple without pegging the thread during the delay. I did get it though. The issues still persist, but I am now better able to handle them.

As a part of figuring this out with you, I updated the docs to make it more clear that the sync calls throw errors. and that they can be handled through the classic `try catch` handling.

I made a PR for this that I hope you will consider merging to master as I think it will help a lot of people with this, and hopefully also prevent anyone from abandoning labjack if they run into this issue.

I put the updates in PR #20 

Once merged, PR #20 should be good enough to close this ticket.
